Consolidate configuration in environment variables

xamanu opened the pull request:
This PR allows all necessary variables for configuration to be defined in the environment variables. It introduces documentation on using an [`.env` file](https://github.com/hotosm/tasking-manager/blob/9095da934ca3ad6f1e8f6813717979059c2418f9/env-example) and updates the docker-compose setup, to start up a development instance with a single command and no need to edit configuration files or fork the code base in any way.

ethan-nelson commented:
Will take a look in a few hours

ethan-nelson commented on server/config.py:
```diff
@@ -23,69 +65,11 @@ class EnvironmentConfig:
+    API_DOCS_URL = f'{APP_BASE_URL}/api-docs/swagger-ui/index.html?' + \
+                   f'url={APP_BASE_URL}/api/docs'
     SEND_FILE_MAX_AGE_DEFAULT = 0
-    SQLALCHEMY_DATABASE_URI = os.getenv('TM_DB', None)
```
should we consider checking for this variable as an alternative to retain backwards compatibility?

ethan-nelson replied:
i guess we have a few other changes on here (e.g. removing the separate environments) so perhaps we just say this will be a point where there are breaking changes and your configuration must be updated in a few ways.

xamanu replied:
Yes, the idea is to have one place where to configure the TM. I hope this will make things easier for people to get it set up.

xamanu replied:
Let's point out a bit more the "breaking" parts.

As far as I see there are two essential things breaking something:

* `TM_DB` - Which is replaced by the `POSTGRES_X` variables to unify with the database configuration (docker container). If wished, we could maintain backwards compatibility. I'd be happy to code the couple of extra lines to handle it.
* `TM_ENV` - Which is becoming deprecated with this PR. It was pushing our very custom configuration for `https://tasks.hotosm.org` and `https://tasks-stage.hotosm.org`. Deprecating this should not effect anybody apart from our main TM instance. Here are `.env` files for both setups that we can use: [tasks.hotosm.org](https://gist.github.com/xamanu/5f30378e4e3294d465463b209a49fce3) and [tasks-stage.hotosm.org](https://gist.github.com/xamanu/9060531ae91109076f4e20b769d88a11).

Is there anything, that I'm missing? @ethan-nelson, @dakotabenjamin and @arunasank, do you prefer to be conservative on the `TM_DB1` variable and maintain backwards compatibility?  I really don't want to make this an annoying change. Just fixing things to get the Docker stuff right, to ease up installation and allow non-devs to test PRs.

ethan-nelson commented on README.md:
````diff
@@ -105,13 +130,17 @@ export TM_SMTP_PORT=smtp-server-port-here
 
 ```bash
-setx TM_DB "postgresql://USER:PASSWORD@HOST/DATABASE"
+setx TM_APP_BASE_URL="web-server-endpoint"
````
the equals sign on the setx commands should be a blank space instead

xamanu replied:
Thanks for spotting! Fixed.

ethan-nelson commented:
Added a few thoughts. I'm not able to test the docker changes locally, but everything else worked fine.

xamanu commented:
Thanks for the review, @ethan-nelson. Fixed some issues based on your feedback and commented on the breaking elements.

xamanu commented:
Together with #1393, this will provide an easy PR testing environment that can be set up with a few console commands, which will be documented in the wiki and/or can be scripted into a helper wrapper.

xamanu commented:
Here are some testing instructions:

* Install a recent version of [Docker](https://docs.docker.com/get-started/) and [Docker Compose](https://docs.docker.com/compose/).
* Check out the code: `git clone https://github.com/hotosm/tasking-manager.git && cd tasking-manager`
* Use a different branch: `git checkout configuration`
* Copy the configuration file `cp env-example .env`
* Add the consumer key and secret in the `.env` file.
* Run the system: `docker-compose up`
* Test whether logging in works under http://127.0.0.1:5000

xamanu commented:
Together with @ramyaragupathy we were testing this PR also in the perspective of new programmer onboarding problems (#1388, #1391). In order to get everything set up, I needed to include PR #1393 into this one.

dakotabenjamin commented:
I tested this after removing all my old docker images and got this error: 
```
Step 13/20 : RUN cd /tmp && npm install
...
> node ./bin/closure-util.js update

info install Downloading http://dl.google.com/closure-compiler/compiler-20171112.zip
info install Downloading http://github.com/google/closure-library/archive/v20170910.zip
info install Download complete: /tmp/node_modules/closure-util/.deps/compiler/5eb2c411b449e9f3d879be0f59139cb5b2894c09
ERR! closure-util getaddrinfo ENOTFOUND codeload.github.com codeload.github.com:443
npm WARN ajv-keywords@3.4.0 requires a peer of ajv@^6.9.1 but none is installed. You must install peer dependencies yourself.
npm WARN acorn-dynamic-import@4.0.0 requires a peer of acorn@^6.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN hot-tasking-manager@3.0.0 No repository field.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.7 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.7: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! closure-util@1.26.0 postinstall: `node ./bin/closure-util.js update`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the closure-util@1.26.0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2019-03-19T18_07_10_260Z-debug.log
ERROR: Service 'migration' failed to build: The command '/bin/sh -c cd /tmp && npm install' returned a non-zero code: 1
```

xamanu commented:
@dakotabenjamin, this error I saw sometimes. Explicitly stating `acorn` and `ajv` will hopefully make this work on your side. Please try again. Thanks.

xamanu commented:
Some further developments here:

* I introduced a backwards capability for `TM_DB` variable.
* Added the functionality that the `tasking-manager.env` is going to be read in, this applies now to the direct setup and the docker one. There isn't any need to set environment variables with `export` or the `.bash_profile` file, anymore. However it still works the old way.

The only **breaking change** is the `TM_ENV`, which is deprecated and the `TM_APP_BASE_URL` that needs to be set.

This should also fix the problems people described in #1388 and #1391.

xamanu commented:
Parts of this became part of the latest release (because of urgency). And the rest I rerolled in #1419 with better description and testing instructions.
